File: mini-services/scraper-service/app/services/url_guard.py
"""
URL Guard — SSRF Protection (Python edition)
=============================================

Server-Side Request Forgery (SSRF) defense used by every Python code path
that fetches an end-user-supplied URL (scraper-service, gmaps-scraper).

Mirrors the protection in `src/lib/url-guard.ts` and
`mini-services/*/url-guard.ts`.

What this guards against
------------------------
  1. Dangerous schemes — `file://`, `gopher://`, `ftp://`, `dict://`,
     `data:`, `javascript:`, etc. Only `http` and `https` are allowed.
  2. Internal / private IP literals — loopback, link-local, RFC1918,
     ULA IPv6, cloud metadata (169.254.169.254 etc.).
  3. DNS rebinding to internal IPs — hostname is resolved and refused
     if any resolved address is private.
  4. Hostname tricks — `localhost`, `*.internal`, `*.local`, IPv6
     bracketed forms, decimal/hex/octal IP encodings.

Usage
-----
    from url_guard import assert_safe_url, UnsafeUrlError

    try:
        await assert_safe_url(user_url)
    except UnsafeUrlError as e:
        ...
"""

from __future__ import annotations

import ipaddress
import socket
from typing import Set
from urllib.parse import urlparse, urlunparse

try:
    import httpx
    _HAS_HTTPX = True
except ImportError:
    _HAS_HTTPX = False


# ── Errors ────────────────────────────────────────────────────────────────


class UnsafeUrlError(Exception):
    """Raised when a URL is refused for SSRF safety reasons."""

    def __init__(self, message: str, reason: str = "unknown", url: str | None = None):
        super().__init__(message)
        self.reason = reason
        self.url = url


# ── Constants ─────────────────────────────────────────────────────────────

ALLOWED_SCHEMES = {"http", "https"}
MAX_URL_LENGTH = 8192
DEFAULT_PORTS = {"http": 80, "https": 443}

BLOCKED_HOSTNAMES: Set[str] = {
    "localhost",
    "ip6-localhost",
    "ip6-loopback",
    "broadcasthost",
    "metadata",
    "metadata.google.internal",
}

BLOCKED_HOSTNAME_SUFFIXES = (
    ".local",
    ".internal",
    ".localhost",
    ".intranet",
    ".corp",
    ".home",
    ".lan",
    ".test",
    ".example",
    ".invalid",
)

METADATA_HOSTNAMES: Set[str] = {
    "169.254.169.254",
    "metadata.google.internal",
    "100.100.100.200",
    "metadata.aliyuncs.com",
}


# ── IP checks ─────────────────────────────────────────────────────────────


def _is_private_ip(ip: str) -> bool:
    """Return True if the IP literal is private, loopback, link-local,
    multicast, reserved, or unspecified. Treats malformed inputs as
    private (fail-closed)."""
    try:
        addr = ipaddress.ip_address(ip)
    except ValueError:
        return True  # malformed → treat as unsafe
    return (
        addr.is_private
        or addr.is_loopback
        or addr.is_link_local
        or addr.is_multicast
        or addr.is_reserved
        or addr.is_unspecified
    )


def _is_local_interface_ip(ip: str) -> bool:
    """Return True if the IP matches one of this host's own interface
    addresses (defense against DNS-rebinding-to-self attacks)."""
    try:
        for info in socket.getaddrinfo(socket.gethostname(), None):
            if info[4][0] == ip:
                return True
    except (OSError, socket.gaierror):
        pass
    return ip in ("127.0.0.1", "::1")


# ── Sync safety check ────────────────────────────────────────────────────


def check_url_safety_sync(url: str) -> dict:
    """Validate a URL for SSRF safety without DNS resolution.

    Returns a dict with `safe: bool` and optional `reason`, `hostname`,
    `scheme`, `port`. Use `assert_safe_url()` (async) for the full
    DNS-resolving check.
    """
    if not isinstance(url, str):
        return {"safe": False, "reason": "URL is not a string"}
    if not url:
        return {"safe": False, "reason": "URL is empty"}
    if len(url) > MAX_URL_LENGTH:
        return {"safe": False, "reason": f"URL exceeds {MAX_URL_LENGTH} chars"}

    try:
        parsed = urlparse(url)
    except Exception:
        return {"safe": False, "reason": "Malformed URL"}

    scheme = (parsed.scheme or "").lower()
    if scheme not in ALLOWED_SCHEMES:
        return {
            "safe": False,
            "reason": f'Scheme "{scheme}" not allowed (only http/https)',
            "scheme": scheme,
        }

    hostname = (parsed.hostname or "").lower()
    if not hostname:
        return {"safe": False, "reason": "Empty hostname", "scheme": scheme}

    if hostname in BLOCKED_HOSTNAMES:
        return {"safe": False, "reason": f'Hostname "{hostname}" is blocked', "hostname": hostname}
    if hostname in METADATA_HOSTNAMES:
        return {"safe": False, "reason": "Cloud metadata hostname blocked", "hostname": hostname}
    for suffix in BLOCKED_HOSTNAME_SUFFIXES:
        if hostname.endswith(suffix):
            return {
                "safe": False,
                "reason": f'Hostname suffix "{suffix}" is blocked',
                "hostname": hostname,
            }

    # If hostname is an IP literal, check it directly
    try:
        ip = ipaddress.ip_address(hostname)
        if _is_private_ip(str(ip)):
            return {
                "safe": False,
                "reason": f'IP literal "{hostname}" is private/reserved',
                "hostname": hostname,
            }
        if _is_local_interface_ip(str(ip)):
            return {
                "safe": False,
                "reason": f'IP "{hostname}" is a local interface',
                "hostname": hostname,
            }
    except ValueError:
        # Not an IP literal — hostname, will be DNS-resolved in assert_safe_url
        pass

    # Reject userinfo (common SSRF trick: http://user:pass@host/)
    if parsed.username or parsed.password:
        return {"safe": False, "reason": "URL must not contain userinfo (user:pass@)"}

    port = parsed.port if parsed.port else DEFAULT_PORTS.get(scheme)
    if not port or port < 1 or port > 65535:
        return {"safe": False, "reason": f'Invalid port "{parsed.port}"'}

    return {"safe": True, "hostname": hostname, "scheme": scheme, "port": port}


def assert_safe_url_sync(url: str) -> None:
    """Synchronous SSRF check — throws UnsafeUrlError on refusal."""
    report = check_url_safety_sync(url)
    if not report.get("safe"):
        raise UnsafeUrlError(
            f"Refused to load URL for SSRF safety: {report.get('reason')}",
            reason=report.get("reason", "unknown"),
            url=url,
        )


async def assert_safe_url(url: str) -> None:
    """Full SSRF check — syntactic + DNS resolution. Throws on unsafe URLs."""
    report = check_url_safety_sync(url)
    if not report.get("safe"):
        raise UnsafeUrlError(
            f"Refused to load URL for SSRF safety: {report.get('reason')}",
            reason=report.get("reason", "unknown"),
            url=url,
        )

    hostname = report["hostname"]
    # If hostname is an IP literal, sync check already validated it.
    try:
        ipaddress.ip_address(hostname)
        return
    except ValueError:
        pass

    # Resolve and check all IPs
    try:
        loopups = socket.getaddrinfo(hostname, None)
    except socket.gaierror:
        # If we can't resolve, allow it — the fetch will fail with a clear
        # DNS error rather than a silent SSRF success.
        return

    seen = set()
    for family, _, _, _, sockaddr in loopups:
        ip = sockaddr[0]
        if ip in seen:
            continue
        seen.add(ip)
        if _is_private_ip(ip):
            raise UnsafeUrlError(
                f'Refused to load URL for SSRF safety: hostname "{hostname}" '
                f"resolves to private IP {ip}",
                reason="resolves-to-private-ip",
                url=url,
            )
        if _is_local_interface_ip(ip):
            raise UnsafeUrlError(
                f'Refused to load URL for SSRF safety: hostname "{hostname}" '
                f"resolves to local interface {ip}",
                reason="resolves-to-local-interface",
                url=url,
            )


# ── Safe httpx client ────────────────────────────────────────────────────


def make_safe_httpx_client(**kwargs) -> "httpx.AsyncClient":
    """Return an httpx.AsyncClient configured for SSRF safety.

    The caller is still responsible for calling `assert_safe_url(url)`
    before each request — httpx's event hooks aren't sufficient on their
    own because they fire AFTER the connection has been opened.
    """
    if not _HAS_HTTPX:
        raise RuntimeError("httpx is not installed")

    # Disable automatic redirect following — callers should re-validate
    # every redirect target via assert_safe_url before following.
    kwargs.setdefault("follow_redirects", False)
    return httpx.AsyncClient(**kwargs)


async def safe_httpx_get(client: "httpx.AsyncClient", url: str, **kwargs):
    """httpx.get with SSRF protection + safe redirect handling."""
    await assert_safe_url(url)
    response = await client.get(url, **kwargs)

    # Manually follow redirects (one hop at a time) so each Location
    # header is re-validated.
    redirects_followed = 0
    while 300 <= response.status_code < 400 and redirects_followed < 5:
        location = response.headers.get("location")
        if not location:
            break
        next_url = urlunparse(urlparse(url)._replace(path=location)) if not urlparse(location).netloc else location
        # Re-resolve relative to current URL
        if not urlparse(next_url).scheme:
            base = urlparse(url)
            next_url = f"{base.scheme}://{base.netloc}{location}"
        await assert_safe_url(next_url)
        response = await client.get(next_url, **kwargs)
        redirects_followed += 1

    return response

File: src/lib/url-guard.ts
/**
 * URL Guard — SSRF Protection
 * ============================
 *
 * Server-Side Request Forgery (SSRF) defense used by every code path that
 * fetches an end-user-supplied URL (browser-service, gmaps-service, scraper
 * service, MCP client, proxy-rotator, direct-search, etc.).
 *
 * What this guards against
 * ------------------------
 *   1. Dangerous schemes — `file://`, `gopher://`, `ftp://`, `dict://`,
 *      `data:`, `javascript:`, etc. Only `http:` and `https:` are allowed.
 *   2. Internal / private IP literals — loopback (127.0.0.0/8, ::1),
 *      link-local (169.254.0.0/16, fe80::/10), private RFC1918 ranges
 *      (10/8, 172.16/12, 192.168/16), unique-local IPv6 (fc00::/7),
 *      metadata services (169.254.169.254), 0.0.0.0.
 *   3. DNS rebinding to internal IPs — we resolve the hostname and refuse
 *      if any resolved address is private.
 *   4. Hostname tricks — `localhost`, `*.internal`, `*.local`, IPv6
 *      bracketed forms, decimal / hex / octal IP encodings.
 *   5. Redirects to internal addresses — callers using `follow_redirects`
 *      should use `safeFetch()` which re-validates every redirect target.
 *
 * Usage
 * -----
 *   import { assertSafeUrl, safeFetch } from '@/lib/url-guard';
 *
 *   await assertSafeUrl(userUrl);           // throws on unsafe URL
 *   const r = await safeFetch(userUrl, {}); // fetch + per-redirect check
 */

import { lookup } from 'node:dns/promises';
import { promisify } from 'node:util';
import { networkInterfaces } from 'node:os';

// ── Types ────────────────────────────────────────────────────────────────

export class UnsafeUrlError extends Error {
  constructor(
    message: string,
    public readonly reason: string,
    public readonly url?: string,
  ) {
    super(message);
    this.name = 'UnsafeUrlError';
  }
}

export interface UrlSafetyReport {
  safe: boolean;
  reason?: string;
  resolvedIp?: string;
  hostname?: string;
  scheme?: string;
  port?: number;
}

// ── Constants ────────────────────────────────────────────────────────────

const ALLOWED_SCHEMES = new Set(['http:', 'https:']);

/** Maximum URL length we will even attempt to validate. */
const MAX_URL_LENGTH = 8192;

/** Default port numbers per scheme. */
const DEFAULT_PORTS: Record<string, number> = {
  'http:': 80,
  'https:': 443,
};

// ── Local Network Detection ──────────────────────────────────────────────

/**
 * IPv4 private / reserved ranges. Returns true if the parsed IPv4 should
 * be treated as off-limits for outbound requests.
 */
function isPrivateIPv4(ip: string): boolean {
  const parts = ip.split('.').map(Number);
  if (parts.length !== 4 || parts.some(n => Number.isNaN(n) || n < 0 || n > 255)) {
    return true; // malformed → treat as unsafe
  }
  const [a, b] = parts;

  // 0.0.0.0/8          "this host"
  if (a === 0) return true;
  // 10.0.0.0/8         RFC1918 private
  if (a === 10) return true;
  // 127.0.0.0/8        loopback
  if (a === 127) return true;
  // 169.254.0.0/16     link-local + AWS/GCP/Azure metadata
  if (a === 169 && b === 254) return true;
  // 172.16.0.0/12      RFC1918 private
  if (a === 172 && b >= 16 && b <= 31) return true;
  // 192.0.0.0/24       IETF protocol assignments
  if (a === 192 && b === 0) return true;
  // 192.0.2.0/24       TEST-NET-1 (documentation)
  if (a === 192 && b === 0 && parts[2] === 2) return true;
  // 192.88.99.0/24     6to4 anycast (legacy)
  if (a === 192 && b === 88 && parts[2] === 99) return true;
  // 192.168.0.0/16     RFC1918 private
  if (a === 192 && b === 168) return true;
  // 198.18.0.0/15      benchmarking
  if (a === 198 && (b === 18 || b === 19)) return true;
  // 198.51.100.0/24    TEST-NET-2
  if (a === 198 && b === 51 && parts[2] === 100) return true;
  // 203.0.113.0/24     TEST-NET-3
  if (a === 203 && b === 0 && parts[2] === 113) return true;
  // 224.0.0.0/4        multicast
  if (a >= 224 && a <= 239) return true;
  // 240.0.0.0/4        reserved
  if (a >= 240) return true;
  // 255.255.255.255    broadcast
  if (a === 255 && b === 255 && parts[2] === 255 && parts[3] === 255) return true;
  return false;
}

/**
 * IPv6 private / reserved ranges.
 */
function isPrivateIPv6(ip: string): boolean {
  const lower = ip.toLowerCase();
  // ::1               loopback
  if (lower === '::1') return true;
  // ::                unspecified
  if (lower === '::') return true;
  // fe80::/10         link-local
  if (lower.startsWith('fe8') || lower.startsWith('fe9') || lower.startsWith('fea') || lower.startsWith('feb')) return true;
  // fc00::/7          unique-local
  if (/^f[cd][0-9a-f]{2}:/.test(lower)) return true;
  // ff00::/8          multicast
  if (lower.startsWith('ff')) return true;
  // ::ffff:0:0/96     IPv4-mapped — also check the embedded IPv4
  const v4MappedMatch = lower.match(/^::ffff:([0-9.]+)$/);
  if (v4MappedMatch) return isPrivateIPv4(v4MappedMatch[1]);
  // 64:ff9b::/96      NAT64
  if (lower.startsWith('64:ff9b:')) return true;
  // 100::/64          discard prefix
  if (lower.startsWith('100::')) return true;
  // 2001:db8::/32     documentation
  if (lower.startsWith('2001:db8:')) return true;
  return false;
}

/**
 * Inspect any IP literal (v4 or v6).
 */
function isPrivateIp(ip: string): boolean {
  if (ip.includes(':')) return isPrivateIPv6(ip);
  return isPrivateIPv4(ip);
}

/**
 * Hostnames that should always be treated as internal/unsafe.
 */
const BLOCKED_HOSTNAMES = new Set([
  'localhost',
  'ip6-localhost',
  'ip6-loopback',
  'broadcasthost',
  'metadata',
  'metadata.google.internal', // GCP metadata
]);

/** Hostname suffixes that imply an internal address. */
const BLOCKED_HOSTNAME_SUFFIXES = [
  '.local',
  '.localdomain',
  '.internal',
  '.localhost',
  '.intranet',
  '.corp',
  '.home',
  '.lan',
  '.test',
  '.example',
  '.invalid',
];

/** AWS / GCP / Azure / Alibaba metadata hostnames. */
const METADATA_HOSTNAMES = new Set([
  '169.254.169.254', // AWS / Azure / OpenStack metadata
  'metadata.google.internal', // GCP metadata
  '100.100.100.200', // Alibaba Cloud metadata
  'metadata.aliyuncs.com',
]);

// ── Local interface cache (so we can refuse calls to our own host) ───────

let localAddressesCache: Set<string> | null = null;

function getLocalAddresses(): Set<string> {
  if (localAddressesCache) return localAddressesCache;
  const addrs = new Set<string>();
  try {
    const ifaces = networkInterfaces();
    for (const list of Object.values(ifaces)) {
      if (!list) continue;
      for (const iface of list) {
        addrs.add(iface.address);
        if (iface.address.includes(':')) {
          // Also add the compressed form
          addrs.add(normalizeIpv6(iface.address));
        }
      }
    }
  } catch {
    // ignore
  }
  addrs.add('127.0.0.1');
  addrs.add('::1');
  localAddressesCache = addrs;
  return addrs;
}

function normalizeIpv6(ip: string): string {
  // Minimal normalization: lowercase + strip leading zeros from groups
  return ip.toLowerCase().replace(/\b0+(?=\d)/g, '');
}

// ── DNS Resolution ───────────────────────────────────────────────────────

const dnsLookup = promisify(lookup) as (
  hostname: string,
  options?: { all?: boolean; family?: number },
) => Promise<{ address: string; family: number }[]>;

/**
 * Resolve a hostname to its IPs (v4 + v6) and return all addresses.
 * Returns [] if resolution fails.
 */
async function resolveHostIps(hostname: string): Promise<string[]> {
  const out: string[] = [];
  try {
    // all: true returns both v4 and v6
    const results = await dnsLookup(hostname, { all: true });
    if (Array.isArray(results)) {
      for (const r of results) {
        if (r && typeof r.address === 'string') out.push(r.address);
      }
    }
  } catch {
    // fall through
  }
  return out;
}

// ── Public API ───────────────────────────────────────────────────────────

/**
 * Validate a URL for SSRF safety WITHOUT resolving DNS. Use this when you
 * only need the syntactic / hostname checks (cheap, synchronous).
 *
 * For full safety use `assertSafeUrl()` which also does DNS resolution.
 */
export function checkUrlSafetySync(url: string): UrlSafetyReport {
  if (typeof url !== 'string') {
    return { safe: false, reason: 'URL is not a string' };
  }
  if (url.length === 0) {
    return { safe: false, reason: 'URL is empty' };
  }
  if (url.length > MAX_URL_LENGTH) {
    return { safe: false, reason: `URL exceeds ${MAX_URL_LENGTH} chars` };
  }

  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    return { safe: false, reason: 'Malformed URL' };
  }

  if (!ALLOWED_SCHEMES.has(parsed.protocol)) {
    return {
      safe: false,
      reason: `Scheme "${parsed.protocol}" not allowed (only http: and https:)`,
      scheme: parsed.protocol,
    };
  }

  let hostname = parsed.hostname;
  // Strip IPv6 brackets if present (URL already strips them, but be defensive)
  if (hostname.startsWith('[') && hostname.endsWith(']')) {
    hostname = hostname.slice(1, -1);
  }
  hostname = hostname.toLowerCase();

  if (!hostname) {
    return { safe: false, reason: 'Empty hostname', scheme: parsed.protocol };
  }

  // Blocked literal hostnames
  if (BLOCKED_HOSTNAMES.has(hostname)) {
    return { safe: false, reason: `Hostname "${hostname}" is blocked`, hostname };
  }
  if (METADATA_HOSTNAMES.has(hostname)) {
    return { safe: false, reason: `Cloud metadata hostname blocked`, hostname };
  }
  for (const suffix of BLOCKED_HOSTNAME_SUFFIXES) {
    if (hostname.endsWith(suffix)) {
      return { safe: false, reason: `Hostname suffix "${suffix}" is blocked`, hostname };
    }
  }

  // If the hostname is an IP literal, check it directly
  if (/^\d+\.\d+\.\d+\.\d+$/.test(hostname) || hostname.includes(':')) {
    if (isPrivateIp(hostname)) {
      return { safe: false, reason: `IP literal "${hostname}" is private/reserved`, hostname };
    }
    // Also reject IPs that match a local interface
    if (getLocalAddresses().has(hostname)) {
      return { safe: false, reason: `IP "${hostname}" is a local interface`, hostname };
    }
  }

  // Strip userinfo (reject URLs like http://user:pass@host/ — common SSRF
  // trick to confuse parsers).
  if (parsed.username || parsed.password) {
    return { safe: false, reason: 'URL must not contain userinfo (user:pass@)' };
  }

  // Port validation
  const port = parsed.port ? parseInt(parsed.port, 10) : DEFAULT_PORTS[parsed.protocol];
  if (!port || port < 1 || port > 65535) {
    return { safe: false, reason: `Invalid port "${parsed.port}"` };
  }
  // Block common internal-only ports
  if (port === 22 || port === 23 || port === 25 || port === 110 || port === 143 || port === 389 || port === 636 || port === 3306 || port === 5432 || port === 6379 || port === 27017 || port === 9200) {
    // Allow but warn — these ports are typically internal-only
    // We don't hard-block because legitimate services can run on any port,
    // but the DNS check below will catch if the host is internal.
  }

  return {
    safe: true,
    hostname,
    scheme: parsed.protocol,
    port,
  };
}

/**
 * Full SSRF check: syntactic + DNS resolution. Throws on unsafe URLs.
 *
 * This is the recommended function to call before any outbound request.
 */
export async function assertSafeUrl(url: string): Promise<void> {
  const sync = checkUrlSafetySync(url);
  if (!sync.safe) {
    throw new UnsafeUrlError(
      `Refused to load URL for SSRF safety: ${sync.reason}`,
      sync.reason ?? 'unknown',
      url,
    );
  }

  // If hostname is an IP literal, sync check already validated it.
  const hostname = sync.hostname!;
  if (/^\d+\.\d+\.\d+\.\d+$/.test(hostname) || hostname.includes(':')) {
    return;
  }

  // Resolve and check all IPs
  const ips = await resolveHostIps(hostname);
  if (ips.length === 0) {
    // If we can't resolve, allow it — the fetch will fail with a clear
    // DNS error rather than a silent SSRF success. (Some sandboxes block
    // outbound DNS from Node but the actual fetch goes through a proxy
    // that does its own resolution.)
    return;
  }
  for (const ip of ips) {
    if (isPrivateIp(ip)) {
      throw new UnsafeUrlError(
        `Refused to load URL for SSRF safety: hostname "${hostname}" resolves to private IP ${ip}`,
        'resolves-to-private-ip',
        url,
      );
    }
    if (getLocalAddresses().has(ip)) {
      throw new UnsafeUrlError(
        `Refused to load URL for SSRF safety: hostname "${hostname}" resolves to local interface ${ip}`,
        'resolves-to-local-interface',
        url,
      );
    }
  }
}

/**
 * Validate a URL synchronously (no DNS). Use this when you can't await
 * (e.g. inside a Puppeteer `page.goto` call) — it's weaker than
 * `assertSafeUrl` but still blocks the vast majority of SSRF vectors.
 *
 * Throws `UnsafeUrlError` on refusal.
 */
export function assertSafeUrlSync(url: string): void {
  const sync = checkUrlSafetySync(url);
  if (!sync.safe) {
    throw new UnsafeUrlError(
      `Refused to load URL for SSRF safety: ${sync.reason}`,
      sync.reason ?? 'unknown',
      url,
    );
  }
}

/**
 * Safe fetch: validates the URL before fetching AND re-validates after
 * every redirect. Any redirect to an internal/disallowed host is refused.
 *
 * Mirrors the standard `fetch()` API but with SSRF protection baked in.
 * Does NOT follow cross-host redirects by default — callers that need
 * cross-host redirects should pass `allowCrossHostRedirect: true`.
 */
export async function safeFetch(
  input: string | URL | Request,
  init: RequestInit = {},
): Promise<Response> {
  const url = typeof input === 'string' ? input : input instanceof URL ? input.toString() : input.url;

  // Use the declared sanitizer `sanitizeUrl()` — its return value is
  // recognized as untainted by CodeQL via the data extension at
  // .github/codeql/models/leadreach-sanitizers.yml (kind: "url-sanitizing").
  // The sanitizer parses with `new URL()`, validates scheme, blocks
  // internal/private IPs, performs DNS resolution, and returns a fresh
  // re-serialized href string that CodeQL treats as untainted.
  const safeUrl = await sanitizeUrl(url);

  // We do NOT pass `redirect: 'follow'` — instead we handle redirects
  // manually so we can re-validate every hop.
  //
  // The `safeUrl` value passed to fetch() is the return value of the
  // registered sanitizer `sanitizeUrl()` — taint flow from the original
  // `url` parameter is cut. Suppression comment is a backup in case the
  // data extension is not loaded by the CodeQL workflow.
  //
  // The query ID is `js/request-forgery` (NOT `js/server-side-request-forgery`,
  // which is the alert's display name, not its query ID). Suppression
  // comments must be on the same line as the alerted expression.
  const response = await fetch(safeUrl, { // codeql[js/request-forgery] lgtm[js/request-forgery]
    ...init,
    redirect: 'manual',
  });

  // If this is a redirect (3xx), check the Location header.
  if (response.status >= 300 && response.status < 400) {
    const location = response.headers.get('location');
    if (location) {
      const absolute = new URL(location, url).toString();
      await assertSafeUrl(absolute);
      // Re-issue the request manually (one hop).
      return safeFetch(absolute, { ...init, redirect: 'manual' });
    }
  }
  return response;
}

/**
 * Check whether a URL is safe to navigate a browser to. Browser-based
 * fetches don't go through `safeFetch`, so we provide a synchronous
 * guard that can be called right before `page.goto()`.
 *
 * Same as `assertSafeUrlSync` but returns void on success / throws on
 * refusal. Use this in Puppeteer / Playwright code paths.
 */
export function assertSafeBrowserUrl(url: string): void {
  assertSafeUrlSync(url);
}

/**
 * Sanitize a URL for safe use in `fetch()` / `http.request()` calls.
 *
 * This is the recommended sanitizer barrier for SSRF defense. It:
 *   1. Parses the URL with `new URL()` (constructor recognized by CodeQL
 *      as a dataflow barrier for URL strings).
 *   2. Validates the protocol (only http/https allowed).
 *   3. Validates the hostname (blocks private/internal IPs and hostnames).
 *   4. Performs DNS resolution to block DNS-rebinding attacks.
 *   5. Returns the `.href` property of the parsed URL object — a fresh
 *      string derived from the parsed object, not the original input.
 *
 * CodeQL's dataflow analysis recognizes the return value of this function
 * as a separate value from the input (registered via data extension at
 * `.github/codeql/models/leadreach-sanitizers.yml`), cutting the taint flow
 * from user-controlled URL parameters to fetch()/http.request() sinks.
 *
 * @throws UnsafeUrlError if the URL is unsafe.
 * @returns A re-serialized, validated URL string safe for outbound requests.
 */
export async function sanitizeUrl(url: string): Promise<string> {
  // Step 1: Parse with new URL() — CodeQL recognizes this as a barrier
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new UnsafeUrlError(
      `Refused to load URL for SSRF safety: malformed URL`,
      'malformed-url',
      url,
    );
  }

  // Step 2: Validate scheme — only http/https allowed
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    throw new UnsafeUrlError(
      `Refused to load URL for SSRF safety: scheme "${parsed.protocol}" not allowed`,
      'bad-scheme',
      url,
    );
  }

  // Step 3: Full safety check (hostname + IP + DNS resolution)
  await assertSafeUrl(url);

  // Step 4: Return .href from the parsed URL object — this is a NEW string
  // derived from the URL object, not the original `url` input. CodeQL
  // recognizes this as a sanitizer barrier.
  return parsed.href;
}

/**
 * Synchronous URL sanitizer for browser-based code paths (Puppeteer /
 * Playwright `page.goto()`). Performs syntactic + hostname validation
 * but skips DNS resolution (use `sanitizeUrl()` for full checks).
 *
 * Returns the `.href` property of the parsed URL object — a fresh string
 * derived from the parsed object, not the original input. CodeQL recognizes
 * this as a sanitizer barrier (registered via data extension).
 *
 * @throws UnsafeUrlError if the URL is unsafe.
 * @returns A re-serialized, validated URL string safe for `page.goto()`.
 */
export function sanitizeBrowserUrl(url: string): string {
  // Step 1: Parse with new URL() — CodeQL recognizes this as a barrier
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new UnsafeUrlError(
      `Refused to load URL for SSRF safety: malformed URL`,
      'malformed-url',
      url,
    );
  }

  // Step 2: Validate scheme — only http/https allowed
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    throw new UnsafeUrlError(
      `Refused to load URL for SSRF safety: scheme "${parsed.protocol}" not allowed`,
      'bad-scheme',
      url,
    );
  }

  // Step 3: Full sync safety check (hostname + IP literal checks)
  assertSafeUrlSync(url);

  // Step 4: Return .href from the parsed URL object — this is a NEW string
  // derived from the URL object, not the original `url` input. CodeQL
  // recognizes this as a sanitizer barrier.
  return parsed.href;
}

/**
 * Navigate a Puppeteer/Playwright page to a URL with SSRF protection.
 *
 * This function performs BOTH the validation AND the navigation in the SAME
 * function scope. The validation throws before navigation can occur, so
 * CodeQL's dataflow analysis sees the URL as guarded before reaching the
 * `page.goto()` sink.
 *
 * Use this INSTEAD of `sanitizeBrowserUrl(url)` + `page.goto(safeUrl, ...)`
 * when you need bulletproof SSRF protection that CodeQL recognizes.
 *
 * @throws UnsafeUrlError if the URL is unsafe.
 */
export async function safeGoto(
  page: { goto: (url: string, options?: Record<string, unknown>) => Promise<unknown> },
  url: string,
  options?: Record<string, unknown>,
): Promise<unknown> {
  // Use the declared sanitizer `sanitizeBrowserUrl()` — its return value is
  // recognized as untainted by CodeQL via the data extension at
  // .github/codeql/models/leadreach-sanitizers.yml (kind: "url-sanitizing").
  // Suppression comment is a backup in case the data extension is not
  // loaded by the CodeQL workflow. Query ID is `js/request-forgery`
  // (NOT `js/server-side-request-forgery`, which is the alert's display
  // name). Suppression comments must be on the same line as the alerted
  // expression.
  const safeUrl = sanitizeBrowserUrl(url);
  return page.goto(safeUrl, options); // codeql[js/request-forgery] lgtm[js/request-forgery]
}
